### PR TITLE
[#992][#1025] Custom Google Play publishing

### LINKED ISSUE #992: GP Deploy Error: Error: PKIX path building failed
### Describe the issue
```
> Task :app:compileZcashmainnetDebugKotlin


FAILURE: Build failed with an exception.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

* What went wrong:
Execution failed for task ':app:processZcashmainnetReleaseVersionCodes'.
> A failure occurred while executing com.github.triplet.gradle.play.tasks.ProcessArtifactVersionCodes$VersionCoder
   > Error getting subject token from metadata server: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
```

### Any extra information that might be useful in the debugging process.
- https://github.com/zcash/secant-android-wallet/issues/935
- https://github.com/Triple-T/gradle-play-publisher/issues/1102

### LINKED ISSUE #1025: Automated deployment with the Google Java API
## Is your feature request related to a problem? Please describe.
We'd like to move away from the https://github.com/Triple-T/gradle-play-publisher. To have a more flexible and future-prone solution, it'd be ideal to adopt automated deployment using Google Java API directly.

It would require changing the version number incrementing to a different strategy, e.g., maybe using a number of git commits as a monotonically increasing number. It still could be queried from the Play Store, but we're thinking about simplicity so that the only API call that has to be made is the upload. We also need to look up JGit, and we have to change the GitHub checkout fetch depth.

## Describe the solution you'd like
Solution working with both a local machine with necessary secrets and the new CI script with customizable variables and automated deployment from our CI.

## Additional context
- https://github.com/zcash/secant-android-wallet/issues/935
- https://github.com/zcash/secant-android-wallet/issues/992
- https://github.com/Triple-T/gradle-play-publisher/issues/1102
- https://developers.google.com/android-publisher
- https://github.com/googleapis/google-api-java-client-services/tree/main/clients/google-api-services-androidpublisher/v3
- https://docs.gradle.org/current/userguide/more_about_tasks.html
- https://developers.google.com/resources/api-libraries/documentation/androidpublisher/v3/java/latest/com/google/api/services/androidpublisher/package-summary.html

### PR DESCRIPTION
**Pull request by HonzaR**
- This implements continuous delivery logic that works from a local machine as well as from our CI scripts. It replaces broken and deprecated Android Publisher Gradle Plugin in favor of custom implementation which is completely under our control.
- There is one follow-up related to Google Play Publish API authorization #1033
- Closes #992
- Closes #1025

> **Note**
>This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] **Self-review** your own code in GitHub's web interface[^1]
- [ ] Add **automated tests** as appropriate
- [x] Update the [**manual tests**](../blob/main/docs/testing/manual_testing)[^2] as appropriate
- [ ] Check the **code coverage**[^3] report for the automated tests
- [x] Update **documentation** as appropriate (e.g [README.md](../blob/main/README.md), and [**Architecture.md**](../blob/main/docs/Architecture.md), etc.)
- [x] **Run the app** and try the changes
- [x] Pull in the latest changes from the **main** branch and **squash** your commits before assigning a reviewer[^4]

> **Note**
> It is good practice to provide before and after UI  **screenshots** in the description of this PR. This is only applicable for changes that modify the UI.

# Reviewer

- [x] Check the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) **checklist**
- [ ] Perform an **ad hoc review**[^5]
- [ ] Review the **automated tests**
- [ ] Review the **manual tests**
- [ ] Review the **documentation**, [**README.md**](../blob/main/README.md), and [**Architecture.md**](../blob/main/docs/Architecture.md) as appropriate
- [ ] **Run the app** and try the changes[^6]

[^1]: _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
[^2]: _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down._
[^3]: _While we are not looking for perfect coverage, the tool can point out potential cases that have been missed. Code coverage can be generated with: `./gradlew check` for Kotlin modules and `./gradlew connectedCheck -PIS_ANDROID_INSTRUMENTATION_TEST_COVERAGE_ENABLED=true` for Android modules._
[^4]: _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._
[^5]: _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
[^6]: _While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

### CODE REVIEW

**Review by LukasKorba — APPROVED** [review]

File: build-conventions-secant/src/main/kotlin/secant.publish-conventions.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+            val ks = KeyStore.getInstance(KeyStore.getDefaultType())
+            FileInputStream(trustStore).use { fis ->
+                ks.load(fis, trustStorePassword?.toCharArray())
+            }
+            NetHttpTransport.Builder().trustCertificates(ks).build()
+        }
+    }
+
+    private fun createHttpTransport(): HttpTransport {
+        val protocols = arrayOf("http", "https")
```
Comment (daira) [inline build-conventions-secant/src/main/kotlin/secant.publish-conventions.gradle.kts:83]:
```suggestion
        val protocols = arrayOf("https", "http")
```
to prefer https if both sets of properties are set.

File: build-conventions-secant/src/main/kotlin/secant.publish-conventions.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+    }
+}
+
+/**
+ * The release track identifier. This class also serves as a type-safe custom task input validation.
+ */
+enum class PublishTrack {
+    INTERNAL,   // Internal testing track
+    ALPHA,  // Closed testing track
+    BETA,   // Opened testing track. Note that use of this track is not supported by this task.
```
Comment (daira) [inline build-conventions-secant/src/main/kotlin/secant.publish-conventions.gradle.kts:286]:
```suggestion
    BETA,   // Open testing track. Note that use of this track is not supported by this task.
```

Comment (daira) [inline build-conventions-secant/src/main/kotlin/secant.publish-conventions.gradle.kts:286]:
```suggestion
    ALPHA,      // Closed testing track
    BETA,       // Open testing track. Note that use of this track is not supported by this task.
```

File: build-conventions-secant/src/main/kotlin/secant.publish-conventions.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+
+    @Throws(IllegalStateException::class)
+    fun toGooglePlayIdentifier(): String {
+        return when (this) {
+            INTERNAL -> "internal"  // $NON-NLS-1$
+            ALPHA -> "alpha"    // $NON-NLS-1$
+            BETA -> error("For security reasons, this script does not support the $this option. Promote the app " +
+                "manually from a lower testing channel instead.")
+            PRODUCTION -> error("For security reasons, this script does not support the $this option. Promote the app" +
+                " manually from a lower testing channel instead.")
```
Comment (daira) [inline build-conventions-secant/src/main/kotlin/secant.publish-conventions.gradle.kts:306]:
```suggestion
            BETA, PRODUCTION -> error("For security reasons, this script does not support the $this option. " +
                "Promote the app manually from a lower testing channel instead.")
```

File: build-conventions-secant/src/main/kotlin/secant.publish-conventions.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+/**
+ * The status of a release. This class also serves as a type-safe custom task input validation.
+ */
+enum class PublishStatus {
+    STATUS_UNSPECIFIED, // Unspecified status.
+    DRAFT,  // The release's APKs are not being served to users.
+    IN_PROGRESS,    // The release's APKs are being served to a fraction of users, determined by 'userFraction'.
+    HALTED, // The release's APKs will no longer be served to users. Users who already have these APKs are unaffected.
+    COMPLETED;  // The release will have no further changes. Its APKs are being served to all users, unless they are
+    // eligible to APKs of a more recent release.
```
Comment (daira) [inline build-conventions-secant/src/main/kotlin/secant.publish-conventions.gradle.kts:320]:
```suggestion
    DRAFT,              // The release's APKs are not being served to users.
    IN_PROGRESS,        // The release's APKs are being served to a fraction of users, determined by 'userFraction'.
    HALTED,             // The release's APKs will no longer be served to users. Users who already have these
                        // APKs are unaffected.
    COMPLETED;          // The release will have no further changes. Its APKs are being served to all users,
                        // unless they are eligible for APKs of a more recent release.
```

File: docs/Deployment.md [hunk 5]
```diff
@@ -14,17 +14,21 @@ After signing is configured, it is possible to then configure deployment to Goog
 ## Automated Deployment
 Automated deployment to Google Play configured with the [Gradle Play Publisher plugin](https://github.com/Triple-T/gradle-play-publisher).
 To perform a deployment:
-1. Configure a Google Cloud service API key with the correct permissions
+1. Configure a Google Cloud service account and API key with the correct permissions
 1. Configure Gradle properties
+    1. `ZCASH_GOOGLE_PLAY_SERVICE_ACCOUNT` - Set the Google Play Service Account enabled in the Google Cloud console
```
Comment (daira) [inline docs/Deployment.md:19]:
```suggestion
    1. `ZCASH_GOOGLE_PLAY_SERVICE_ACCOUNT` - Enable the Google Play Service Account in the Google Cloud console
```

File: docs/Deployment.md [hunk 6]
```diff
@@ -14,17 +14,21 @@ After signing is configured, it is possible to then configure deployment to Goog
...
-    1. `ZCASH_GOOGLE_PLAY_DEPLOY_MODE` - Set to `deploy`
+    1. `ZCASH_GOOGLE_PLAY_DEPLOY_TRACK` - Set to `internal`
+    1. `ZCASH_GOOGLE_PLAY_DEPLOY_STATUS` - Set to `completed`
 1. Run the Gradle task `./gradlew :app:publishBundle`
 
 To generate a build with a correct version that can be deployed manually later:
-1. Configure a Google Cloud service API key with the correct permissions
+1. Configure a Google Cloud service account and API key with the correct permissions
 1. Configure Gradle properties
+    1. `ZCASH_GOOGLE_PLAY_SERVICE_ACCOUNT` - Set the Google Play Service Account enabled in the Google Cloud console
```
Comment (daira) [inline docs/Deployment.md:28]:
```suggestion
    1. `ZCASH_GOOGLE_PLAY_SERVICE_ACCOUNT` - Set to the Google Play Service Account that is enabled in the Google Cloud console
```

**Review by daira — APPROVED** [review]
utACK with non-blocking suggestions.

File: build-conventions-secant/src/main/kotlin/secant.publish-conventions.gradle.kts (hunk 4 above)
Comment (HonzaR) [reply build-conventions-secant/src/main/kotlin/secant.publish-conventions.gradle.kts:320]: Unfortunately, this is how [Ktlint](https://github.com/pinterest/ktlint) formats and requires comments to be used.

File: docs/Deployment.md (hunk 5 above)
Comment (HonzaR) [reply docs/Deployment.md]: This is correct wording, as we want user to set the Google Play Service Account enabled in the Google Cloud console to this gradle field.

**Pull request merged**